bmap-rs/src/main.rs

```diff
@@ -87,41 +101,100 @@ fn setup_input(path: &Path) -> Result<Decoder> {
+    let bmap = match input_type {
+        InputType::Local => {
+            find_bmap(&c.image).ok_or_else(|| anyhow!("Couldn't find bmap file"))?
+        }
+        InputType::Remote => find_bmap(Path::new(&c.image.file_name().unwrap()))
+            .ok_or_else(|| anyhow!("Couldn't find bmap file"))?,
+    };
     println!("Found bmap file: {}", bmap.display());
 
     let mut b = File::open(&bmap).context("Failed to open bmap file")?;
     let mut xml = String::new();
     b.read_to_string(&mut xml)?;
 
     let bmap = Bmap::from_xml(&xml)?;
-    let mut output = std::fs::OpenOptions::new()
-        .write(true)
-        .create(true)
-        .open(c.dest)?;
-
-    if output.metadata()?.is_file() {
-        ftruncate(output.as_raw_fd(), bmap.image_size() as i64)
-            .context("Failed to truncate file")?;
+    let output = match input_type {
+        InputType::Local => {
+            let file = std::fs::OpenOptions::new()
+                .write(true)
+                .create(true)
+                .open(c.dest)?;
+            Output::Sync(file)
+        }
+        InputType::Remote => {
+            let file = tokio::fs::OpenOptions::new()
+                .write(true)
+                .create(true)
+                .open(c.dest)
+                .await?;
+            Output::Async(file)
+        }
+    };
+
+    match output {
+        Output::Sync(ref output) => {
+            if output.metadata()?.is_file() {
+                ftruncate(output.as_raw_fd(), bmap.image_size() as i64)
+                    .context("Failed to truncate file")?;
+            }
+        }
+        Output::Async(ref output) => {
+            if output.metadata().await?.is_file() {
+                ftruncate(output.as_raw_fd(), bmap.image_size() as i64)
+                    .context("Failed to truncate file")?;
+            }
+        }
     }
 
-    let mut input = setup_input(&c.image)?;
-    bmap::copy(&mut input, &mut output, &bmap)?;
-    println!("Done: Syncing...");
-    output.sync_all().expect("Sync failure");
+    match output {
+        Output::Sync(mut output) => {
+            let mut input = setup_local_input(&c.image)?;
+            bmap::copy(&mut input, &mut output, &bmap)?;
+            println!("Done: Syncing...");
+            output.sync_all().expect("Sync failure");
+        }
+        Output::Async(mut output) => {
+            let res = setup_remote_input(&c.image).await;
+            let stream = res
+                .bytes_stream()
+                .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))
+                .into_stream();
+            let read = StreamReader::new(stream);
+            let reader = async_compression::tokio::bufread::GzipDecoder::new(read);
+            let mut input = AsyncDiscarder::new(reader);
+            bmap::copy_async(&mut input, &mut output, &bmap).await?;
+            println!("Done: Syncing...");
```

Comment:
duplicate code

Reply:
Yes, I'm trying to find a way to reduce duplicated code but the fact that the output has a different type for sync (std::fs::File) and async (tokio:fs:File) make me always duplicate code. I'm trying to find a workaround like when I've created an enumerate "Output" to hold both types inside.